src/Middleware/RequestAutoPopulationMiddleware.php

```diff
@@ -0,0 +1,92 @@
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+namespace Google\ApiCore\Middleware;
+
+use Google\ApiCore\Call;
+use GuzzleHttp\Promise\PromiseInterface;
+use Ramsey\Uuid\Uuid;
+
+/**
+ * Middleware that adds autopopulation functionality. This middlware is
+ * added iff auto population settings are present in the resource
+ * descriptor config for the rpc method in context.
+ *
+ * @internal
+ */
+class RequestAutoPopulationMiddleware implements MiddlewareInterface
+{
+    /** @var callable */
+    private $nextHandler;
+
+    /** @var array<string, string> */
+    private $autoPopulationSettings;
+
+    public function __construct(
+        callable $nextHandler,
+        array $autoPopulationSettings
+    ) {
+        $this->nextHandler = $nextHandler;
+        $this->autoPopulationSettings = $autoPopulationSettings;
+    }
+
+    /**
+     * @param Call $call
+     * @param array $options
+     *
+     * @return PromiseInterface
+     */
+    public function __invoke(Call $call, array $options)
+    {
+        $next = $this->nextHandler;
+
+        $request = $call->getMessage();
+        foreach ($this->autoPopulationSettings as $fieldName => $valueType) {
+            $getFieldName = 'get' . ucwords($fieldName);
+            // Populate the field if it's not already set by user
+            if (empty($request->$getFieldName())) {
+                $setFieldName = 'set' . ucwords($fieldName);
+                switch ($valueType) {
+                    case \Google\Api\FieldInfo\Format::UUID4:
+                        $request->$setFieldName(Uuid::uuid4()->toString());
+                        break;
+                    default:
+                        // Do nothing
+                }
+            }
+        }
+        $call = $call->withMessage($request);
```

Comment:
We probably want to wrap this in `if ($this-autoPopulationSettings)` so that executing this with empty settings is a more efficient no-op

Reply:
I've opened a PR showing this refactoring https://github.com/googleapis/gax-php/pull/545.

Reply:
Merged